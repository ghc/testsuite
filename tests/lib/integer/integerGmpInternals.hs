{-# LANGUAGE MagicHash, UnboxedTuples #-}

module Main (main) where

import GHC.Base
import GHC.Integer
import GHC.Integer.GMP.Internals

gcdExtInteger' :: Integer -> Integer -> (Integer, Integer)
gcdExtInteger' a b = case gcdExtInteger a b of (# a, b #) -> (a,b)

powInteger' :: Integer -> Word -> Integer
powInteger' b (W# w#) = powInteger b w#

{- Reference implementation for 'powModInteger'

powModIntegerHs :: Integer -> Integer -> Integer -> Integer
powModIntegerHs b0 e0 m
  | e0 >= 0    = go b0 e0 1
  | otherwise  = error "non-neg exponent required"
  where
    go !b e !r
      | odd e     = go b' e' (r*b `mod` m)
      | e == 0    = r
      | otherwise = go b' e' r
      where
        b' = b*b `mod` m
        e' = e   `unsafeShiftR` 1 -- slightly faster than "e `div` 2"

-}

main :: IO ()
main = do
    print $ powModInteger b e m
    print $ gcdExtInteger' b e
    print $ gcdExtInteger' e b
    print $ gcdExtInteger' x y
    print $ gcdExtInteger' y x
    print $ powInteger' 12345 0
    print $ powInteger' 12345 1
    print $ powInteger' 12345 30
    print $ [ (x,i) | x <- [0..71], let i = recipModInteger x (2*3*11*11*17*17), i /= 0 ]
    return ()
  where
    b = 2988348162058574136915891421498819466320163312926952423791023078876139
    e = 2351399303373464486466122544523690094744975233415544072992656881240319
    m = 10^(40::Int)

    x = 5328841272400314897981163497728751426
    y = 32052182750761975518649228050096851724
