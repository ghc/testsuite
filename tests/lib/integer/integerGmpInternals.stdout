1527229998585248450016808958343740453059
682382427572745901624116300491295556924
682382427572745901624116300491295556924
(1,-238164827888328100873319793437342927637138278785737103723156342382925)
(1,302679100340807588460107986194035692812415103244388831792688023418704)
(92889294,115110207004456909698806038261)
(92889294,-19137667681784054624628973533)
1
12345
555562377826831043419246079513769804614412256811161773362797946971665712715296306339052301636736176350153982639312744140625
[(1,1),(5,41963),(7,59947),(13,177535),(19,143557),(23,182447),(25,134281),(29,7235),(31,33841),(35,95915),(37,113413),(41,61409),(43,24397),(47,174101),(49,158431),(53,193979),(59,188477),(61,185737),(65,35507),(67,118999),(71,186173)]
2988348162058574136915891421498819466320163312926952423791023078876343
2351399303373464486466122544523690094744975233415544072992656881240451
[2,3,5,7,11,13,17,19,23,29,31,37,41,43,47,53,59,61,67,71,73,79,83,89,97,101,103,107,109,113,127,131,137,139,149,151,157,163,167,173,179,181,191,193,197,199]
[(25,0),(1,1),(261,0),(1,1),(107,0),(1,1),(49,0),(1,1),(27,0),(1,1),(137,0),(1,1),(49,0),(1,1),(53,0),(1,1),(29,0),(1,1),(39,0),(1,1),(89,0),(1,1),(37,0),(1,1),(21,0),(1,1),(65,0)]
[(132,0),(1,1),(75,0),(1,1),(551,0),(1,1),(240,0)]
